### ISSUE TITLE
Inconsistency in rewrited code

### ISSUE THREAD

**Issue by AlexanderDKazakov** [RamanathanVaradharajan/Namics#49]
- I figured out that the commit aca888ef9b4ec7ffbbd6c64c7f6cd0ebf7cfcf9c reduces logic in procedure checkInput() in lattice.cpp:

![screen shot 2018-12-22 at 19 35 01](https://user-images.githubusercontent.com/17707394/50377612-bfe9c700-0620-11e9-8ce8-5060026216d5.png)

These removed and added parts have different logic.  New logic leads to broken bondary conditions.
As a result all time we have only mirror boundary condition.

- There is another new behaviour in namics: if I move clamped node more than one lattice point I get seg fault after some time. I still could not find the place where exactly it happens, but debugger told me the last point where the program runs is collectphi().

I don't know what will be better either go to commit before aca888ef9b4ec7ffbbd6c64c7f6cd0ebf7cfcf9c or go through the code again.

**Comment by Dmry** [comment]
Hi sasha! The idea behind this is the following: The if statement collects the string that's set in the input file for boundary conditions, and if it's not found it defaults to mirror. PutM is called at the end of the check_input function, which sets the boundary conditions according to the file. The removed part was redundant. I'll double check the implementation and update you later today!

Did you check valgrind to see which memory location gives the segfault? Alternatively you can send me an input file and I'll see what I can find.

**Comment by Dmry** [comment]
Sorry about the delay. You're correct about the broken boundaries. Working on a fix now.

**Comment by Dmry** [comment]
Should now be fixed in the master branch by Pull request #50. @AlexanderDKazakov , can you check whether this also maybe happens to fix you segfault?

**Comment by AlexanderDKazakov** [comment]
Great! Thank you for your help! I will check and reply!

**Comment by AlexanderDKazakov** [comment]
I am sorry for delay. Here is my input file: 

```
lat : flat : gradients : 3
lat : flat : n_layers_x : 20
lat : flat : n_layers_y : 20
lat : flat : n_layers_z : 20
lat : flat : geometry : planar
lat : flat : lattice_type : simple_cubic
lat : flat : lowerbound_x : periodic
lat : flat : upperbound_x : periodic
lat : flat : lowerbound_y : periodic
lat : flat : upperbound_y : periodic
lat : flat : lowerbound_z : periodic
lat : flat : upperbound_z : periodic

mon : W : freedom : free
mon : A : freedom : free
mon : X : freedom : clamp
mon : X : clamp_info : (5, 5, 5)(10, 10, 10);(10, 10, 10)(15, 15, 5);(10, 10, 10)(5, 15, 15);(10, 10, 10)(15, 5, 15);(15, 15, 5)(20, 20, 10);(20, 20, 10)(25, 25, 5);(20, 20, 10)(25, 15, 15);(20, 20, 10)(15, 25, 15);(5, 15, 15)(10, 20, 20);(10, 20, 20)(5, 25, 25);(10, 20, 20)(15, 15, 25);(10, 20, 20)(15, 25, 15);(15, 5, 15)(20, 10, 20);(20, 10, 20)(25, 5, 25);(20, 10, 20)(15, 15, 25);(20, 10, 20)(25, 15, 15);
mon : X : sub_box_size : 18

mol : pol : freedom : clamped
mol : pol : composition : (X)1(A)48(X)1
mol : water : freedom : solvent
mol : water : composition : (W)1

newton : isaac : method : DIIS
newton : isaac : m : 4
newton : isaac : deltamax : 0.1
newton : isaac : e_info : true
newton : isaac : iterationlimit : 7000
newton : isaac : tolerance : 1e-3

out_info : folder : path : data/

output : vtk : append : false
vtk : mol : pol : phi

output : kal : append : true
kal : output : varyable : MC_step
kal : sys : varyable : free_energy

output : pro : append : false
pro : mon : X : phi

output : pos : append : true
pos : output : noname : X
pos : output : noname : Y
pos : output : noname : Z

output : vec : append : true
vec : mol : pol : gn
vec : output : noname : gn

output : ana : append : false


cleng : parameters : MCS : 3000
cleng : parameters : save_interval : 1
cleng : parameters : delta_step : 2
cleng : parameters : checkpoint_load : true
cleng : parameters : checkpoint_save : true
cleng : parameters : cleng_pos : true

start
```

I varied delta_step and sometimes it happened that I got a segfault. If only it reproducing more frequently. Currently, I can't reproduce it. If you succeed in it, it would be great.

**Comment by AlexanderDKazakov** [comment]
Ha! I found it! This input file will show the same.  After crossing PBC all my nodes go not in the same position as before. I will send you animation how it works on your email. Maybe this behavior comes from cleng.

**Comment by Dmry** [comment]
Great! I'll have a look at it anyway though, see if I can find any other inconsistencies. Let's leave this issue open until we have an actual fix for the problem.

**Comment by AlexanderDKazakov** [comment]
Hmm.. I checked it, everything is up-today. I will look again.

**Comment by AlexanderDKazakov** [comment]
I looked again:

Initially, I fixed MC step and set it equal all time = {0, 0, 2}
We have only 2 nodes and they are connected between each other. Each MC step cleng choose one of them and make shift using MC step.

So system looks like:
![same_shift_box](https://user-images.githubusercontent.com/17707394/51316193-31683a00-1a54-11e9-9e34-f46acd6e6f8a.png)

I expect that nodes will cross PBC and nothing strange will happen with free energy.

What I got:
![same_shift](https://user-images.githubusercontent.com/17707394/51316328-a20f5680-1a54-11e9-9cf0-b8db947245eb.png)

I don't know why at very beginning 2 picks slightly higher than others.
I understand why free energy so high at range from 300 to 500. It is because both nodes are on different sides of PBC.

And after time at MC step 800 I got segfault. I noticed that Namics remember positions of node and don't shift it to trivial cell (e.g id: 0 { 23, 23, 829 }; id: 1 { 29, 29, 829 }; )

I looked at the output which I can visualize: I saw a drift of nodes position. Both nodes drifted to a vertical direction (it is the same as I sent you). However it seems that Namics don't do it, maybe it produced by writing output. 


I looked at the output from a debugger. What I got:

![image](https://user-images.githubusercontent.com/17707394/51315619-d5e97c80-1a52-11e9-9d92-4660ce1e098b.png)

```
Signal = SIGSEGV (Segmentation fault)
k = {int} 12
j = {int} 1
i = {int} 1
p = {int} 0
phi = {double * | 0x7efcceda2c48} 0x7efcceda2c48
GN = {Real * | 0x5597248e93d0} 0x5597248e93d0
rho = {Real * | 0x5597248feb10} 0x5597248feb10
Bx = {int * | 0x559724667980} 0x559724667980
By = {int * | 0x5597246679a0} 0x5597246679a0
Bz = {int * | 0x5597246679c0} 0x5597246679c0
MM = {int} 12167
M = {int} 2744
n_box = {int} 1
Mx = {int} 12
My = {int} 12
Mz = {int} 12
MX = {int} 21
MY = {int} 21
MZ = {int} 21
jx = {int} 196
jy = {int} 14
JX = {int} 529
JY = {int} 23
pos_l = {int} 0
pos_x = {int} 11109
pos_y = {int} 483
pos_z = {int} 815
Bxp = {int} 20
Byp = {int} 20
Bzp = {int} 824
Inv_H_GNp = {Real} 142464943.15962571
ii = {int} 196
jj = {int} 14
kk = {int} 12
```

It is interesting why it is so.
